### ISSUE TITLE
Defined constants in a namespaced file are not registered

### ISSUE THREAD

**Issue by hkdobrev** [maglnet/ComposerRequireChecker#64]
I have the following file defining constants:

```php
<?php

namespace Foo;

define('BAR', 'baz');
```

This code works in PHP and uses the built-in `define` function.

However, these lines:

https://github.com/maglnet/ComposerRequireChecker/blob/281435afb02468595ee139694f2befd5f5ca2bf6/src/ComposerRequireChecker/NodeVisitor/DefinedSymbolCollector.php#L90-L95

make the checker ignore this pattern and the code needs to change to:

```php
<?php

namespace Foo;

\define('BAR', 'baz');
```

for the constant to be recognised.

The review comments in https://github.com/maglnet/ComposerRequireChecker/pull/59#discussion_r173026361 are fair and I think the case where the `define()` function is defined in a namespace needs to tackled, but the check should be different to allow for the more common case of not defining a `define()` function in a namespace, but just using the built-in one without a leading slash.

**Comment by hkdobrev** [comment]
I got another case which works with the checker:

```php
<?php

namespace Foo;

use function define;

define('BAR', 'baz');
```

The use of `\define()` works with `composer-require-checker`, but not with PHP Code Sniffer about PSR-1. And the `use function define;` works with both `composer-require-checker` and `phpcs`, but is removed from our rules for PHP CS Fixer as obsolete.

Could you please point me in the right direction where the issue is as I think both checks in `composer-require-checker` and PHPCS are not handling common cases. See for example [this commit](https://github.com/squizlabs/PHP_CodeSniffer/commit/5f4c9cc7145df818f8edf352b13bfc17a9a623b8) for how PHPCS detects it.

**Comment by hkdobrev** [comment]
I'll try to clarify on your points:

- I am using the built-in `define()` function.
- I want to register a constant which is not in a namespace.
- I do that from a file which has a namespace, though.
- I need to go through hoops to do that.
- PSR-1 is not obsolete (PSR-0 is). We actually enforce PSR-2, but it includes PSR-1 in itself and PHPCS reports the violation as part of PSR-1.

The specific violation is:

```shell
$ phpcs --standard=PSR2 foo.php
FILE: foo.php
-----------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------------------
 1 | WARNING | A file should declare new symbols (classes, functions, constants, etc.)
   |         | and cause no other side effects, or it should execute logic with side
   |         | effects, but should not do both. The first symbol is defined on line 5
   |         | and the first side effect is on line 5.
-----------------------------------------------------------------------------------------
```

I think the above is a problem in PHPCS as the `define()` function actually works and is the built-in one.

However, `composer-require-checker` is not considering the `define()` function to be the default either.

My solution in the end is to move constant definitions in a file which is not in a namespace.

So do you think these tools should detect I am using the built-in `define()` in these cases or is it just my problem I am doing it from a namespaced file?

**Comment by Ocramius** [comment]
> * I need to go through hoops to do that.

**Comment by hkdobrev** [comment]
OK, thanks! 😸

**Comment by Ocramius** [comment]
> So do you think these tools should detect I am using the built-in `define()` in these cases or is it just my problem I am doing it from a namespaced file?

Yes, this is clearly a phpcs issue :-) (sorry, closed the issue by accident)

**Comment by Ocramius** [comment]
Btw, the `doctrine/coding-standard` forces `use function` and doesn't have these issues, so I think you might be running on some older sniffs...

**Comment by hkdobrev** [comment]
@Ocramius Both code blocks you've posted above work in `composer-require-checker`, but not on the latest PHPCS. I was indeed on an older version and just upgraded to the latest stable one. Reading their code, doesn't seem to be supported.

But fair enough, if these 2 work in the require checker, I think it's fine for static code analyser to not check whether I have a `define()` function in that namespace.

I'll just report to PHPCS the case with `\define()` which is not supported at the moment and hopefully we'll get to the bottom of this.

My issue with `use function` is only due to the rules in PHP CS Fixer.
